### PR TITLE
Add [ThreadStatic] correctness analyzers

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/46626

Note that we'd discussed adding a fixer for the case of `[ThreadStatic]` applied to an instance field, but on further thought I don't think that's a good idea.  If someone is using `[ThreadStatic]` on an instance field, it's possible it's a mistake and they intended for it to be static, but it's also quite likely they assumed it would create a per-thread instance value (as you might get with `ThreadLocal<T>`) so providing a fix to just mark it static would potentially lead someone down the wrong path, and just adding `static` is trivial for someone to do.  There should also be very few violations of this, so a bulk fix isn't going to be meaningful.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx [hunk 1]
```diff
@@ -1449,6 +1449,24 @@
   <data name="UseEnvironmentCurrentManagedThreadIdFix" xml:space="preserve">
     <value>Use 'Environment.CurrentManagedThreadId'</value>
   </data>
+  <data name="ThreadStaticOnNonStaticFieldDescription" xml:space="preserve">
+    <value>'ThreadStatic' only affects static fields.  When applied to instance fields, it has no impact on behavior.</value>
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:1453]: This looks like an extra space

```suggestion
    <value>'ThreadStatic' only affects static fields. When applied to instance fields, it has no impact on behavior.</value>
```

**Review by Youssef1313 — COMMENTED** [review]
Edge cases to consider whether they should be supported:

```csharp
using System;

public class C
{
    [field: ThreadStatic]
    public string S { get; }
}

public record R([field: ThreadStatic] string S);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    var fieldInit = (IFieldInitializerOperation)context.Operation;
+                    foreach (IFieldSymbol field in fieldInit.InitializedFields)
+                    {
+                        if (field.IsStatic && field.HasAttribute(threadStaticAttributeType))
+                        {
+                            context.ReportDiagnostic(fieldInit.CreateDiagnostic(ThreadStaticInitializedInlineRule));
+                        }
+                    }
+
+                }, OperationKind.FieldInitializer);
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs:78]: The loop here will probably result to unnecessary duplicate diagnostics. Consider checking one field only. I *think* all fields will have the same value for `IsStatic` and `.HasAttribute()`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx (hunk 1 above)
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx]: I removed the space, as I assume from your comment two-spaces-after-a-period is not the style used in analyzers.  It's just what I've always done :)

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs (hunk 2 above)
Comment (stephentoub) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs]: I changed it to exit the loop after the first diagnostic.

**Comment by stephentoub** [comment]
> Edge cases to consider whether they should be supported:

I added support for properties (which implicitly adds support for records).  I don't believe it's possible today to find the backing field for an IEventSymbol.

**Comment by stephentoub** [comment]
Thanks for reviewing, @Youssef1313.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    // If it's a static symbol, nothing to see here.
+                    if (symbol.IsStatic)
+                    {
+                        return;
+                    }
+
+                    // If it's a field, it's the symbol we'll check.
+                    // If it's an auto-prop, find its backing field if there is one.
+                    IFieldSymbol? fieldSymbol = symbol as IFieldSymbol;
+                    if (symbol is IPropertySymbol)
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs:70]: We have an extension method to determine if a symbol is a property with backing field: https://github.com/dotnet/roslyn-analyzers/blob/2b89a4115019ef1280706a6631b6b7c8f9840c9a/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L110-L114

Consider adding an out `[NotNullWhen(true)] out IFieldSymbol? backingField` parameter to it and use it here.

**Review by mavasani — APPROVED** [review]

**Pull request merged**